app/tests/src/db/test_bulk_ops.py

```diff
@@ -0,0 +1,94 @@
+"""Tests for bulk_ops module"""
+import operator
+import random
+from dataclasses import dataclass
+
+from psycopg import rows, sql
+
+import src.adapters.db as db
+from src.db import bulk_ops
+
+
+@dataclass
+class Number:
+    id: str
+    num: int
+
+
+def get_random_number_object() -> Number:
+    return Number(
+        id=str(random.randint(1000000, 9999999)),
+        num=random.randint(1, 10000),
+    )
+
+
+def test_bulk_upsert(db_session: db.Session):
+    db_client = db.PostgresDBClient()
+    conn = db_client.get_raw_connection()
+    # conn = db_session.connection().connection
+    # Override mypy, because SQLAlchemy's DBAPICursor type doesn't specify the row_factory attribute, or that it functions as a context manager
+    with conn.cursor(row_factory=rows.class_row(Number)) as cur:  # type: ignore
+        table = "temp_table"
+        attributes = ["id", "num"]
+        objects = [get_random_number_object() for i in range(100)]
+        constraint = "temp_table_pkey"
+
+        # Create a table for testing bulk upsert
+        cur.execute(
+            sql.SQL(
+                "CREATE TEMP TABLE {table}"
+                "("
+                "id TEXT NOT NULL,"
+                "num INT,"
+                "CONSTRAINT {constraint} PRIMARY KEY (id)"
+                ")"
+            ).format(
+                table=sql.Identifier(table),
+                constraint=sql.Identifier(constraint),
+            )
+        )
+
+        bulk_ops.bulk_upsert(
+            cur,
+            table,
+            attributes,
+            objects,
+            constraint,
+        )
+        conn.commit()
+
+        # Check that all the objects were inserted
+        cur.execute(
+            sql.SQL("SELECT id, num FROM {table} ORDER BY id ASC").format(
+                table=sql.Identifier(table)
+            )
+        )
+        records = cur.fetchall()
+        objects.sort(key=operator.attrgetter("id"))
+        assert records == objects
+
+        # Now modify half of the objects
+        for obj in objects[: int(len(objects) / 2)]:
+            obj.num = random.randint(1, 10000)
+
+        # And insert additional objects
+        objects.extend([get_random_number_object() for i in range(50)])
+
+        bulk_ops.bulk_upsert(
+            cur,
+            table,
+            attributes,
+            objects,
+            constraint,
+        )
+        conn.commit()
```

Comment:
nit: issue i have with this testis that even though you only modify half of the objects, you're still doing an update on all the objects since you're doing upsert with the full object list. i think we should only pass the modifed objects to the upsert call and make sure the old objects are still there.

put another way, i could introduce a bug without breaking this test i.e. if i replaced the upsert implementation with a "truncate then insert" that will still work, since the current test passes in all the old unmodified objects.

to make the test cleaner, i recommend separating out the variables you are using to track the expected state of the db and the variables you are using as input parameters. right now you're using `objects` for both input and expected values. similarly, other than the expected state variable which we expect to change, consider treating all variables as constant / don't modify them. in this vein, consider using namedtuples instead of dataclasses which are immutable.

then some optional nice-to-haves, might be nice to pull random objects rather than just the first half of them using random.sample, and mix the updates and inserts.

here's some pseudocode that addresses these ideas:

```
objects_by_id = {}

objects_to_insert = [get_random_number_object() for i in range(100)]
bulk_upsert(..., objects_to_insert, ...)

for obj in objects_to_insert:
  objects_by_id[obj.id] = obj
expected_records = sorted(objects_by_id.values, key=attrgetter("id"))

assert fetch_all_records() == expected_records

objects_to_update = [obj._replace(num=random.randint(1,10000)) for obj in random.sample(objects_by_id.values(), 50)]
more_objects_to_insert = [get_random_number_object() for i in range(50)]
objects_to_upsert = objects_to_update + more_objects_to_insert
random.shuffle(objects_to_upsert)

bulk_upsert(..., objects_to_upsert, ...)

for obj in objects_to_upsert:
  objects_by_id[obj.id] = obj
expected_records = sorted(objects_by_id.values, key=attrgetter("id"))

assert fetch_all_records() == expected_records
```

Reply:
👍 makes sense! I made these changes